core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1702,21 +1727,28 @@ abstract class RDD[T: ClassTag](
       persist(LocalRDDCheckpointData.transformStorageLevel(storageLevel), allowOverride = true)
     }
 
-    // If this RDD is already checkpointed and materialized, its lineage is already truncated.
-    // We must not override our `checkpointData` in this case because it is needed to recover
-    // the checkpointed data. If it is overridden, next time materializing on this RDD will
-    // cause error.
-    if (isCheckpointedAndMaterialized) {
-      logWarning("Not marking RDD for local checkpoint because it was already " +
-        "checkpointed and materialized")
-    } else {
-      // Lineage is not truncated yet, so just override any existing checkpoint data with ours
-      checkpointData match {
-        case Some(_: ReliableRDDCheckpointData[_]) => logWarning(
-          "RDD was already marked for reliable checkpointing: overriding with local checkpoint.")
-        case _ =>
-      }
+
+    RDDCheckpointData.synchronized {
```

Comment:
Holding this lock while calling `doCheckpoint` causes deadlock. So this code was changed to hold the lock only during the needed period.